vm/systemSmartContracts/delegation.go

```diff
@@ -1026,7 +1080,147 @@ func (d *delegation) unDelegate(args *vmcommon.ContractCallInput) vmcommon.Retur
+	}
+	if len(args.Arguments) != 0 {
+		d.eei.AddReturnMessage("must call without arguments")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) < 0 {
+		d.eei.AddReturnMessage("cannot call with negative value")
+		return vmcommon.UserError
+	}
+
+	totalActiveData := d.eei.GetStorage([]byte(totalActiveKey))
+	serviceFeeData := d.eei.GetStorage([]byte(serviceFeeKey))
+	rewardsData := &RewardComputationData{
+		RewardsToDistribute: args.CallValue,
+		TotalActive:         big.NewInt(0).SetBytes(totalActiveData),
+		ServiceFee:          big.NewInt(0).SetBytes(serviceFeeData).Uint64(),
+	}
+	currentEpoch := d.eei.BlockChainHook().CurrentEpoch()
+	err := d.saveRewardData(currentEpoch, rewardsData)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.UserError
+}
+
+func (d *delegation) getRewardData(epoch uint32) (bool, *RewardComputationData, error) {
+	marshaledData := d.eei.GetStorage(rewardKeyForEpoch(epoch))
+	if len(marshaledData) == 0 {
+		return false, nil, nil
+	}
+	rewardsData := &RewardComputationData{}
+	err := d.marshalizer.Unmarshal(rewardsData, marshaledData)
+	if err != nil {
+		return false, nil, err
+	}
+
+	return true, rewardsData, nil
+}
+
+func rewardKeyForEpoch(epoch uint32) []byte {
+	epochInBytes := big.NewInt(int64(epoch)).Bytes()
+	return append([]byte(rewardKeyPrefix), epochInBytes...)
+}
+
+func (d *delegation) saveRewardData(epoch uint32, rewardsData *RewardComputationData) error {
+	marshaledData, err := d.marshalizer.Marshal(rewardsData)
+	if err != nil {
+		return err
+	}
+
+	d.eei.SetStorage(rewardKeyForEpoch(epoch), marshaledData)
+	return nil
+}
+
+func (d *delegation) calculateAndUpdateRewards(callerAddress []byte, delegator *DelegatorData) error {
+	activeFund, err := d.getFund(delegator.ActiveFund)
+	if err != nil {
+		return err
+	}
+
+	isOwner := d.isOwner(callerAddress)
+
+	totalRewards := big.NewInt(0)
+	currentEpoch := d.eei.BlockChainHook().CurrentEpoch()
+	for i := delegator.RewardsCheckpoint; i <= currentEpoch; i++ {
+		found, rewardData, errGet := d.getRewardData(i)
+		if errGet != nil {
+			return errGet
+		}
+		if !found {
+			continue
+		}
+
+		percentage := float64(rewardData.ServiceFee) / float64(percentageDenominator)
+		rewardsForOwner := core.GetPercentageOfValue(rewardData.RewardsToDistribute, percentage)
+		rewardForDelegator := big.NewInt(0).Sub(rewardData.RewardsToDistribute, rewardsForOwner)
+
+		// delegator reward is: rewardForDelegator * user stake / total delegation cap
+		rewardForDelegator.Mul(rewardForDelegator, activeFund.Value)
+		remaining := big.NewInt(0)
+		rewardForDelegator.DivMod(rewardForDelegator, rewardData.TotalActive, remaining)
+
+		if isOwner {
+			totalRewards.Add(totalRewards, rewardsForOwner)
+			totalRewards.Add(totalRewards, remaining)
+		}
+
+		totalRewards.Add(totalRewards, rewardForDelegator)
+	}
+
+	delegator.UnClaimedRewards.Add(delegator.UnClaimedRewards, totalRewards)
```

Comment:
do not forget to also add the computed remaining for the owner, otherwise it won't be distributed

Reply:
fixed

Reply:
Is this done?

Reply:
yes. for now dust cannot be redistributed. will think about it.